cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractProgramController.java

```diff
@@ -181,7 +181,9 @@ public void run() {
    * Children call this method to signal the program is completed.
    */
   protected void complete() {
-    if (!state.compareAndSet(State.ALIVE, State.COMPLETED)) {
+    if (!state.compareAndSet(State.STARTING, State.COMPLETED)
```

Comment:
this has a potential race condition if the state changes between the three compareAndSet() calls.The correct way would be to:
```
while (attempts-- > 0) {
  current = state.get();
  if (state not in [alive, starting, suspended] {
    LOG.warn(...);
    fail;
  }
  if (state.compareAndSet(current, complete)) {
    return;
  } 
}
fail("too many concurrent state changes");